### PR TITLE
CAT: add validation for state messages

### PR DESCRIPTION
**Pull request by lazebnyi**
## What
Fixed https://github.com/airbytehq/airbyte-internal-issues/issues/6557

Emit final state message for full refresh syncs and consolidate read flows was added - https://github.com/airbytehq/airbyte/pull/35622
We need to test that the legacy state type is not emitted and that the source stats are present in the state message

## How

add validation for state messages

## Recommended reading order
1. `airbyte-integrations/bases/connector-acceptance-test/connector_acceptance_test/tests/test_core.py`

## 🚨 User Impact 🚨

no breaking changes

## Pre-merge Actions

<details><summary><strong>Updating the Python CDK</strong></summary>

### Airbyter

Before merging:
- Pull Request description explains what problem it is solving
- Code change is unit tested
- Build and my-py check pass
- Smoke test the change on at least one affected connector
   - On Github: Run [this workflow](https://github.com/airbytehq/airbyte/actions/workflows/connectors_tests.yml), passing `--use-local-cdk --name=source-<connector>` as options
   - Locally: `airbyte-ci connectors --use-local-cdk --name=source-<connector> test`
- PR is reviewed and approved
      
After merging:
- [Publish the CDK](https://github.com/airbytehq/airbyte/actions/workflows/publish-cdk-command-manually.yml)
   - The CDK does not follow proper semantic versioning. Choose minor if this the change has significant user impact or is a breaking change. Choose patch otherwise.
   - Write a thoughtful changelog message so we know what was updated.
- Merge the platform PR that was auto-created for updating the Connector Builder's CDK version
   - This step is optional if the change does not affect the connector builder or declarative connectors.

</details>

### CODE REVIEW

File: airbyte-integrations/bases/connector-acceptance-test/connector_acceptance_test/tests/test_core.py [hunk 1]
```diff
@@ -1254,6 +1259,22 @@ def _validate_stream_statuses(configured_catalog: ConfiguredAirbyteCatalog, stat
             assert status_list[-1] == AirbyteStreamStatus.COMPLETE
             assert all(x == AirbyteStreamStatus.RUNNING for x in status_list[1:-1])
 
+    @staticmethod
+    def _validate_state_messages(state_messages: List[AirbyteMessage]):
+        for state_message in state_messages:
```
Comment (brianjlai) [inline airbyte-integrations/bases/connector-acceptance-test/connector_acceptance_test/tests/test_core.py:1264]: Let's add an additional assertion that there is at least one state message. As part of https://github.com/airbytehq/airbyte-internal-issues/issues/2925 all syncs that end in success should have emitted at least one state message

File: airbyte-integrations/bases/connector-acceptance-test/connector_acceptance_test/tests/test_core.py [hunk 2]
```diff
@@ -1254,6 +1259,22 @@ def _validate_stream_statuses(configured_catalog: ConfiguredAirbyteCatalog, stat
...
+            state = state_message.state
+            stream_name = state.stream.stream_descriptor.name
+            state_type = state.type
+
+            # Ensure legacy state type is not emitted anymore
+            assert state_type != AirbyteStateType.LEGACY, (
```
Comment (brianjlai) [inline airbyte-integrations/bases/connector-acceptance-test/connector_acceptance_test/tests/test_core.py:1270]: 🎉

**Review by brianjlai — COMMENTED** [review]
Overall looks good but I have a couple comments on the validations we have and I think adding a `bypass_reason` like we do on some of our other tests might be good escape hatch. The default should be on however.

File: airbyte-integrations/bases/connector-acceptance-test/connector_acceptance_test/tests/test_core.py (hunk 2 above)
Comment (lazebnyi) [reply airbyte-integrations/bases/connector-acceptance-test/connector_acceptance_test/tests/test_core.py:1270]: 🙂

File: airbyte-integrations/bases/connector-acceptance-test/unit_tests/test_core.py [hunk 3]
```diff
@@ -1699,3 +1704,99 @@ async def test_all_supported_file_types_present(mocker, file_types_found, should
             await t.test_all_supported_file_types_present(certified_file_based_connector=True, inputs=config)
     else:
         await t.test_all_supported_file_types_present(certified_file_based_connector=True, inputs=config)
+
+@pytest.mark.parametrize(
+    ("state_message_params", "should_fail"),
+    (
+            ({"type": AirbyteStateType.STREAM, "sourceStats": AirbyteStateStats(recordCount=1.0)}, False),
+            ({"type": AirbyteStateType.LEGACY, "sourceStats": AirbyteStateStats(recordCount=1.0)}, True),
```
[the remark this answers is not in the document]
Comment (lazebnyi) [reply airbyte-integrations/bases/connector-acceptance-test/unit_tests/test_core.py]: Deleted

File: airbyte-integrations/bases/connector-acceptance-test/connector_acceptance_test/tests/test_core.py (hunk 1 above)
Comment (lazebnyi) [reply airbyte-integrations/bases/connector-acceptance-test/connector_acceptance_test/tests/test_core.py:1275]: Added an additional assertion

**Review by lazebnyi — COMMENTED** [review]
Added the `should_validate_state_messages` flag to disable the test when necessary. 
Decide not to add `bypass_reason` to avoid unnecessary complicity in configuration of this test.

**Review by brianjlai — APPROVED** [review]
lgtm! thanks for making the additions

**Pull request merged**